### PR TITLE
- MLHR-1908 Added helper method to deserialize JSONObject into a Map.

### PR DESCRIPTION
**Pull request by ilooner**

### CODE REVIEW

**Comment by ilooner** [comment]
@brightchen please review

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 1]
```diff
@@ -123,6 +123,54 @@ public static GPOMutable deserialize(FieldsDescriptor fieldsDescriptor,
...
+        throw new IllegalStateException("This should never happen", ex);
+      }
+
+      if (keyValue instanceof JSONArray) {
+
+        JSONArray ja = (JSONArray) keyValue;
+
+        for (int index = 0; index < ja.length(); index++) {
+          Object val = getFieldFromJSON(type, ja, index);
+          keyValues.add(val);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: why not keyValues.add(getFieldFromJSON(type, ja, index));

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 2]
```diff
@@ -123,6 +123,54 @@ public static GPOMutable deserialize(FieldsDescriptor fieldsDescriptor,
...
+        }
+
+      } else if (keyValue instanceof JSONObject) {
+        throw new UnsupportedOperationException("Cannot extract objects from JSONObjects");
+      } else {
+        Object val = getFieldFromJSON(fieldsDescriptor, key, dpou);
+        keyValues.add(val);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: same.. why not keyValues.add(getFieldFromJSON(fieldsDescriptor, key, dpou));

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 3]
```diff
@@ -123,6 +123,54 @@ public static GPOMutable deserialize(FieldsDescriptor fieldsDescriptor,
...
+                                                          JSONObject dpou)
+  {
+    Map<String, Set<Object>> keyToValues = Maps.newHashMap();
+
+    for (String key : fieldsDescriptor.getFields().getFields()) {
+      if (!dpou.has(key)) {
+        throw new IllegalArgumentException("The given key " + key + " is not contained in the given JSON");
+      }
+
+      Type type = fieldsDescriptor.getType(key);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Can this go into if (keyValue instanceof JSONArray) {} loop

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 4]
```diff
@@ -304,52 +355,216 @@ else if(type == Type.CHAR) {
...
+   * @param index The index of the object in the {@link JSONArray} to retrieve.
+   * @return The object retrieved from the {@link JSONArray}.
+   */
+  public static Object getFieldFromJSON(Type type, JSONArray ja, int index)
+  {
+    if (type == Type.BOOLEAN) {
+      Boolean val;
+
+      try {
+        val = ja.getBoolean(index);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: return ja.getBoolean(index);

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 5]
```diff
@@ -304,52 +355,216 @@ else if(type == Type.CHAR) {
...
+                                           + val
+                                           + " which is too large to fit into a short.");
+      }
+
+      return ((short)val);
+    } else if (type == Type.INTEGER) {
+      int val;
+
+      try {
+        val = ja.getInt(index);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: return ja.getInt(index);

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 6]
```diff
@@ -304,52 +355,216 @@ else if(type == Type.CHAR) {
...
+                                           + " does not have a valid int value.",
+                                           ex);
+      }
+
+      return val;
+    } else if (type == Type.LONG) {
+      long val;
+
+      try {
+        val = ja.getLong(index);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: same return ja.getLong(index);

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 7]
```diff
@@ -304,52 +355,216 @@ else if(type == Type.CHAR) {
...
+                                           + val
+                                           + " that is not one character long.");
+      }
+
+      return val.charAt(0);
+    } else if (type == Type.STRING) {
+      String val;
+
+      try {
+        val = ja.getString(index);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: return ja.getString(index);

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 8]
```diff
@@ -304,52 +355,216 @@ else if(type == Type.CHAR) {
...
+                                           + " does not have a valid string value.",
+                                           ex);
+      }
+
+      return val;
+    } else if (type == Type.DOUBLE) {
+      Double val;
+
+      try {
+        val = ja.getDouble(index);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: same..

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 9]
```diff
@@ -304,52 +355,216 @@ else if(type == Type.CHAR) {
...
+                                           + " does not have a valid double value.",
+                                           ex);
+      }
+
+      return val;
+    } else if (type == Type.FLOAT) {
+      Float val;
+
+      try {
+        val = (float) ja.getDouble(index);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: same as above

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 10]
```diff
@@ -123,6 +123,54 @@ public static GPOMutable deserialize(FieldsDescriptor fieldsDescriptor,
...
+  {
+    Map<String, Set<Object>> keyToValues = Maps.newHashMap();
+
+    for (String key : fieldsDescriptor.getFields().getFields()) {
+      if (!dpou.has(key)) {
+        throw new IllegalArgumentException("The given key " + key + " is not contained in the given JSON");
+      }
+
+      Type type = fieldsDescriptor.getType(key);
+      Set<Object> keyValues = Sets.newHashSet();
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: I think you should not initialize it here. 
In if (keyValue instanceof JSONArray){} loop , initialized it with length of the array and in final else loop initialize it with length 1 using Sets.newHashSetWithExpectedSize(). 

keyToValues.put(key, keyValues) can be the last step in main for loop

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 11]
```diff
@@ -178,33 +226,44 @@ public static void setFieldFromJSON(GPOMutable gpo, Type type, String field, JSO
...
 
-    if(type == Type.BOOLEAN) {
+    if (type == Type.BOOLEAN) {
       Boolean val;
 
       try {
         val = jo.getBoolean(field);
-      }
-      catch(JSONException ex) {
+      } catch (JSONException ex) {
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: why not return jo.getBoolean(field); 

This also should be done for LONG, INTEGER, DOUBLE etc

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 12]
```diff
@@ -123,6 +123,54 @@ public static GPOMutable deserialize(FieldsDescriptor fieldsDescriptor,
...
+
+      Type type = fieldsDescriptor.getType(key);
+      Set<Object> keyValues = Sets.newHashSet();
+      keyToValues.put(key, keyValues);
+
+      Object keyValue;
+
+      try {
+        keyValue = dpou.get(key);
+      } catch (JSONException ex) {
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java:146]: why catch exception? Let the user decide what to do with it

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 13]
```diff
@@ -178,33 +226,44 @@ public static void setFieldFromJSON(GPOMutable gpo, Type type, String field, JSO
...
-    else if(type == Type.BYTE) {
+      return val;
+    } else if (type == Type.BYTE) {
       int val;
 
       try {
         val = jo.getInt(field);
-      }
-      catch(JSONException ex) {
+      } catch (JSONException ex) {
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java:250]: why catch exception?

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 14]
```diff
@@ -304,52 +355,216 @@ else if(type == Type.CHAR) {
...
+        throw new IllegalArgumentException("The index " + index + " does not have a valid bool value.", ex);
+      }
+
+      return val;
+    } else if (type == Type.BYTE) {
+      int val;
+
+      try {
+        val = ja.getInt(index);
+      } catch (JSONException ex) {
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java:358]: why catch exception?

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 1 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 2 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 3 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 4 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 5 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 6 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 7 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 8 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 9 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 10 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 11 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: Done

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 12 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java:146]: This is a convenience method for handling JSON. it is not convenient to do exception handling every time this method is called. It also doesn't enable you to do anything useful.

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 13 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java:250]: Same as above.

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 14 above)
Comment (ilooner) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java:358]: Same as above.

**Comment by ilooner** [comment]
Applied gaurav's comments, @brightchen please do final review.

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 15]
```diff
@@ -178,33 +227,40 @@ public static void setFieldFromJSON(GPOMutable gpo, Type type, String field, JSO
...
-      }
-      catch(JSONException ex) {
+      } catch (JSONException ex) {
         throw new IllegalArgumentException("The key "
                                            + field
                                            + " does not have a valid byte value.", ex);
       }
 
-      if(val < (int)Byte.MIN_VALUE) {
+      if (val < (int)Byte.MIN_VALUE) {
```
Comment (brightchen) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java]: what about combine these two logic into one and throw IllegalArgumentException with "Out of range".

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java [hunk 16]
```diff
@@ -178,33 +227,40 @@ public static void setFieldFromJSON(GPOMutable gpo, Type type, String field, JSO
...
-      Boolean val;
+  /**
+   * This method gets the given field from the given {@link JSONObject} and converts the field to an object
+   * of the type specified in the given {@link FieldsDescriptor}.
+   * @param fd The {@link FieldsDescriptor} describing the type of each field.
+   * @param field The field to retrieve from the given {@link JSONObject}.
+   * @param jo The {@link JSONObject} to retrieve a field from.
+   * @return The value of the given field converted to an object of the correct type.
+   */
+  public static Object getFieldFromJSON(FieldsDescriptor fd, String field, JSONObject jo)
```
Comment (brightchen) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java:242]: maybe can combine the logic of integer numeric( byte, short, int ). such as
if(isIntegerNumeric(type))
{
  int val = jo.getInt(field);
  if( !insideRange(val, type) )
     throw new IllegalArgumentException( "value {} out of range of type {}" );
}

**Comment by ilooner** [comment]
@brightchen Implemented comments

File: library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java (hunk 12 above)
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/appdata/gpo/GPOUtils.java:146]: This is util class and it should not catch exceptions. It should be left to user to swallow it or rethrow it. Now with your changes user doesn't have that control.

**Pull request merged**